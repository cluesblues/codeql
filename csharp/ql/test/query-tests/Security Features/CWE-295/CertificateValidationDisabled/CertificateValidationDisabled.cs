// semmle-extractor-options: /r:System.Security.Cryptography.X509Certificates.dll /r:System.Net.dll /r:System.Net.Http.dll /r:System.Net.Security.dll /r:System.Linq.dll /r:System.Net.Primitives.dll
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading.Tasks;

namespace System.Net
{
    static class ServicePointManager
    {
        public static RemoteCertificateValidationCallback ServerCertificateValidationCallback { get; set; }
    }
}

namespace CertificateValidationDisabled
{
    class Program
    {
        static void Main(string[] args)
        {
            // BAD
            ServicePointManager.ServerCertificateValidationCallback += (sender, cert, chain, sslPolicyErrors) => { return true; };
            ServicePointManager.ServerCertificateValidationCallback += (a, b, c, d) => { return true; };
            ServicePointManager.ServerCertificateValidationCallback += (a, b, c, d) => true;
            ServicePointManager.ServerCertificateValidationCallback += delegate { return true; };

            // BAD
            ServicePointManager.ServerCertificateValidationCallback = new RemoteCertificateValidationCallback(AcceptAllCertifications);

            var m = new Program();
            // BAD
            ServicePointManager.ServerCertificateValidationCallback = new RemoteCertificateValidationCallback(m.AcceptAllCertificationsNonStatic);
            m.Test();

            // GOOD
            ServicePointManager.ServerCertificateValidationCallback += (a, b, c, d) =>
            {
                int x = 20;
                if (x < 100)
                {
                    return false;
                }
                else
                {
                    return true;
                }
            };

            // GOOD
            ServicePointManager.ServerCertificateValidationCallback += delegate (
                object sender,
                X509Certificate cert,
                X509Chain chain,
                SslPolicyErrors sslPolicyErrors)
            {
                if (cert == null)
                {
                    return false;
                }

                if (sslPolicyErrors != SslPolicyErrors.None)
                {
                    return false;
                }

                return true;
            };
        }

        void Test()
        {
            // BAD
            ServicePointManager.ServerCertificateValidationCallback = new RemoteCertificateValidationCallback(this.AcceptAllCertificationsNonStatic);
        }

        public static bool AcceptAllCertifications(object sender, X509Certificate certification, X509Chain chain, SslPolicyErrors sslPolicyErrors)
        {
            return true;
        }

        public bool AcceptAllCertificationsNonStatic(object sender, X509Certificate certification, X509Chain chain, SslPolicyErrors sslPolicyErrors)
        {
            return true;
        }
    }
}
