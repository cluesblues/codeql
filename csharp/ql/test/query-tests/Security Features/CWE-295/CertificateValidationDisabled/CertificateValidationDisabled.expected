edges
| CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... | CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... |
| CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... | CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... |
| CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... | CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... |
| CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } | CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } |
| CertificateValidationDisabled.cs:34:111:34:133 | access to method AcceptAllCertifications | CertificateValidationDisabled.cs:34:71:34:134 | delegate creation of type RemoteCertificateValidationCallback |
| CertificateValidationDisabled.cs:38:111:38:144 | access to method AcceptAllCertificationsNonStatic | CertificateValidationDisabled.cs:38:71:38:145 | delegate creation of type RemoteCertificateValidationCallback |
| CertificateValidationDisabled.cs:42:72:53:13 | (...) => ... | CertificateValidationDisabled.cs:42:72:53:13 | (...) => ... |
| CertificateValidationDisabled.cs:56:72:73:13 | delegate(...) { ... } | CertificateValidationDisabled.cs:56:72:73:13 | delegate(...) { ... } |
| CertificateValidationDisabled.cs:79:111:79:147 | access to method AcceptAllCertificationsNonStatic | CertificateValidationDisabled.cs:79:71:79:148 | delegate creation of type RemoteCertificateValidationCallback |
#select
| CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... | CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... | CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:28:72:28:129 | (...) => ... | this |
| CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... | CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... | CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:29:72:29:103 | (...) => ... | this |
| CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... | CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... | CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:30:72:30:91 | (...) => ... | this |
| CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } | CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } | CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:31:72:31:96 | delegate(...) { ... } | this |
| CertificateValidationDisabled.cs:34:71:34:134 | delegate creation of type RemoteCertificateValidationCallback | CertificateValidationDisabled.cs:34:111:34:133 | access to method AcceptAllCertifications | CertificateValidationDisabled.cs:34:71:34:134 | delegate creation of type RemoteCertificateValidationCallback | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:82:28:82:50 | AcceptAllCertifications | this |
| CertificateValidationDisabled.cs:38:71:38:145 | delegate creation of type RemoteCertificateValidationCallback | CertificateValidationDisabled.cs:38:111:38:144 | access to method AcceptAllCertificationsNonStatic | CertificateValidationDisabled.cs:38:71:38:145 | delegate creation of type RemoteCertificateValidationCallback | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:87:21:87:52 | AcceptAllCertificationsNonStatic | this |
| CertificateValidationDisabled.cs:79:71:79:148 | delegate creation of type RemoteCertificateValidationCallback | CertificateValidationDisabled.cs:79:111:79:147 | access to method AcceptAllCertificationsNonStatic | CertificateValidationDisabled.cs:79:71:79:148 | delegate creation of type RemoteCertificateValidationCallback | Certificate validation disabled for $@ callback. | CertificateValidationDisabled.cs:87:21:87:52 | AcceptAllCertificationsNonStatic | this |
